optuna/trial.py

```diff
@@ -519,42 +537,65 @@ def __init__(self, params):
         # type: (Dict[str, Any]) -> None
 
         self._params = params
+        self._suggested_params = {}  # type: Dict[str, Any]
+        self._distributions = {}  # type: Dict[str, BaseDistribution]
         self._user_attrs = {}  # type: Dict[str, Any]
         self._system_attrs = {}  # type: Dict[str, Any]
 
     def suggest_uniform(self, name, low, high):
         # type: (str, float, float) -> float
 
-        return self._suggest(name)
+        return self._suggest(name, distributions.UniformDistribution(low=low, high=high))
 
     def suggest_loguniform(self, name, low, high):
         # type: (str, float, float) -> float
 
-        return self._suggest(name)
+        return self._suggest(name, distributions.LogUniformDistribution(low=low, high=high))
 
     def suggest_discrete_uniform(self, name, low, high, q):
         # type: (str, float, float, float) -> float
 
-        return self._suggest(name)
+        r = high - low
+
+        if math.fmod(r, q) != 0:
+            high = (r // q) * q + low
+            logger = logging.get_logger(__name__)
+            logger.warning('The range of parameter `{}` is not divisible by `q`, and is '
+                           'replaced by [{}, {}].'.format(name, low, high))
```

Comment:
The logic to restrict the `high` value seems to be duplicated with [L235-L240](https://github.com/pfnet/optuna/blob/add-trial-distributions/optuna/trial.py#L235).
I think we can extract them into a (protected) method of `BaseTrial` to keep maintainability.
What do you think?

Reply:
Sounds good. Thank you for your suggestion.

Reply:
Fixed: https://github.com/pfnet/optuna/pull/383/commits/18515515f1140c936c1ef66b477af997b3f9dbdd

Since this logic does not depend on `self`, I factored out it into a private function.